python/ray/tune/tuner.py

```diff
@@ -0,0 +1,176 @@
+# The motivations behind the new Tuner API as part of Ray MLC effort are that
+# 1. Users can seamlessly transition from their training process to tuning process.
+# 2. Dataset and data preprocessing are now an integral part of training/tuning process.
+#   Bonus point: you can even tune dataset and preprocessing.
+
+# Some requirements on Tuner that may help you understand some implementation decisions:
+# 1. Tuner needs to work in ray client mode.
+# 2. Tuner needs to be restored in case of any failures.
+########################################################################
+
+# The magic key that is used when instantiating Tuner during resume.
+_TUNER_INTERNAL = "tuner_internal"
+_SELF = "self"
+
+
+@PublicAPI(stability="beta")
+class Tuner:
+    """The external facing Tuner class as part of Ray MLC effort.
+
+    Args:
+        trainable (Union[
+                str,
+                Callable,
+                Type[Trainable],
+                Type[ConvertibleToTrainable],
+                ConvertibleToTrainable,
+            ]): The trainable to be tune.
+        param_space (Optional[Dict[str, Any]]): Search space to run tuning on.
+            One thing to note is that both preprocessor and dataset can be tuned here.
+        tune_algo_config (Optional[TuneAlgoConfig]): tune algo specific configs.
+            Refer to ray.tune.tuner_config.TunerAlgoConfig for more info.
+        name (Optional[str]): Name of the run.
+        local_dir (Optional[str]): Local dir to save training results to.
+            Defaults to ``~/ray_results``.
+        callbacks (Optional[List[Callback]]): Callbacks to invoke.
+            Refer to ray.tune.callback.Callback for more info.
+
+    Usage pattern:
+    .. code-block:: python
+
+    # Only do the following, if you want to run in ray client mode.
+    # This means the tune driver code will be running on head node of your cluster.
+    ray.init("ray://my_ray_cluster_address")
+
+    param_space = {
+        "scaling_config": {
+            "num_actors": tune.grid_search([2, 4]),
+            "cpus_per_actor": 2,
+            "gpus_per_actor": 0,
+        },
+        "preprocessor": tune.grid_search([prep_v1, prep_v2]),
+        "datasets": {
+            "train_dataset": tune.grid_search([ds1, ds2]),
+        },
+        "params": {
+            "objective": "binary:logistic",
+            "tree_method": "approx",
+            "eval_metric": ["logloss", "error"],
+            "eta": tune.loguniform(1e-4, 1e-1),
+            "subsample": tune.uniform(0.5, 1.0),
+            "max_depth": tune.randint(1, 9),
+        },
+    }
+    tuner = Tuner(trainable=trainer, param_space=param_space, name="my_tune_run")
+    analysis = tuner.fit()
+
+    This returns an ``ExperimentAnalysis`` object, that you can interact with according
+    to its API.
+
+    If the run finishes in error, to retry
+    you can then do
+    .. code-block:: python
+
+    tuner = Tuner.restore(experiment_checkpoint_dir)
```

Comment:
I'm not entirely convinced we need another checkpoint dir. Are we sure we can't solve this another way? Especially given that this results in yet another argument to `tune.run(...)`.

Reply:
also, this example should run. Import `Tuner` and if there's no checkpoint dir, create one.

Reply:
Just as a note, the `experiment_checkpoint_dir` is still the same here, i.e. we only have one dir containing any kind of checkpointed data. Currently `tune.run()` constructs the experiment checkpoint dir from the experiment name and the local dir, and since we use tune.run as the backend we need to be able to set this explicitly and resolve it on the head node (when using Ray client). Hope this makes sense as an explanation.

TLDR: No new directory, just an argument to explicitly overwrite the directory path if needed